cli/flox/src/commands/environment.rs

```diff
@@ -1362,20 +1368,79 @@ impl Pull {
 
         let pointer_content =
             serde_json::to_string_pretty(&pointer).context("Could not serialize pointer")?;
-        let pointer_path = dot_flox_path.join(ENVIRONMENT_POINTER_FILENAME);
 
-        fs::create_dir_all(&dot_flox_path).context("Could not create .flox/ directory")?;
+        let dot_flox_parent = dot_flox_path
+            .parent()
+            .context("Could not get .flox/ parent")?;
+        fs::create_dir_all(dot_flox_parent).context("Could not create .flox/ parent directory")?;
+
+        let temp_dot_flox_dir = tempfile::TempDir::with_prefix_in(DOT_FLOX, dot_flox_parent)
+            .context("Could not create temporary directory for cloning environment")?;
+
+        let pointer_path = temp_dot_flox_dir.path().join(ENVIRONMENT_POINTER_FILENAME);
         fs::write(pointer_path, pointer_content).context("Could not write pointer")?;
 
-        let result =
-            ManagedEnvironment::open(flox, pointer, &dot_flox_path).map_err(Self::convert_error);
+        let mut env = {
+            let result = Dialog {
+                message,
+                help_message: None,
+                typed: Spinner::new(|| ManagedEnvironment::open(flox, pointer, &temp_dot_flox_dir)),
+            }
+            .spin()
+            .map_err(Self::convert_error);
+
+            match result {
+                Err(err) => {
+                    fs::remove_dir_all(&temp_dot_flox_dir)
+                        .context("Could not clean up .flox/ directory")?;
+                    Err(err)?
+                },
+                Ok(env) => env,
+            }
+        };
+
+        let result = Dialog {
+            message,
+            help_message: None,
+            typed: Spinner::new(|| env.build(flox)),
+        }
+        .spin();
+
         match result {
-            Err(err) => {
-                fs::remove_dir_all(dot_flox_path).context("Could not clean up .flox/ directory")?;
-                Err(err)?;
+            Ok(_) => {
+                fs::rename(temp_dot_flox_dir, dot_flox_path)
+                    .context("Could not move .flox/ directory")?;
+            },
+            Err(
+                e @ EnvironmentError2::Core(CoreEnvironmentError::LockedManifest(
+                    LockedManifestError::BuildEnv(CallPkgDbError::PkgDbError(PkgDbError {
+                        exit_code: 100,
+                        ..
+                    })),
+                )),
+            ) => {
+                if !add_systems && !query_add_system(&flox.system) {
+                    bail!(formatdoc! {"
+                        This environment is not yet compatible with your system ({system}).
+
+                        {err:#}
```

Comment:
suggestion: catch a more specific error (or just look in the manifest.lock) and don't print a build error to the user.

nonblocking suggestion:
I think it's confusing that internally we use the presence of packages to determine whether an environment supports a system rather than looking at options.system

```
> diff /tmp/{1,2}
35d34
<     "aarch64-darwin": {}
> "/nix/store/vjdqnznlwpqx28cr8vfdkjb9bzsb6svv-flox-pkgdb-0.1.0/bin/pkgdb" "buildenv" "$(cat /tmp/1)"
{"store_path":"/nix/store/66csh502qj36s2q78xrzq0w6pqq5nx5k-environment"}
> "/nix/store/vjdqnznlwpqx28cr8vfdkjb9bzsb6svv-flox-pkgdb-0.1.0/bin/pkgdb" "buildenv" "$(cat /tmp/2)"
general error: No packages found for system `aarch64-darwin'
```

I think it would be better to throw a specific exception in pkgdb when a system isn't present in `options.system`, and you can catch that error code here

Reply:
As far as i could tell `"aarch64-darwin": {}` and `"aarch64-darwin": ε` are different cases to pkgdb already.
`"aarch64-darwin": {}` is produced iff the manifest has `"aarch64-darwin ∈ options.system` and means "there is **no package** for aarch64-darwin"
when there is no "aarch64-darwin" in `options.system` then "there is no pacakge **for aarch64-darwin**"
Kinda subtle difference which can't really be told from the message itself.
In fact the error code returned is a `FLOX_EXCEPTION` type, so indeed rather general.
I'll add a more specific one for that case.

I would not like to make this a fail condition of resolution/locking

Reply:
Yeah it's fine, hence nonblocking. But it was confusing/surprising to me that we're relying on that subtle difference (`"aarch64-darwin": {}` vs `"aarch64-darwin": ε`) which is almost accidental/coincidental instead of just relying on the obvious `"aarch64-darwin ∈ options.system`

Reply:
I guess we'd need to check the current error case anyway since we're accessing a an object that may not exist -- sure we can add a test if the system is in `options.systems` not sure if we need to block on that

Reply:
I don't think we need to block on that, just a note. I do think we should block on not printing the error though